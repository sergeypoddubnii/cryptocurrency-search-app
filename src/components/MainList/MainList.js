import React, { useEffect, useMemo } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import mainCryptoOperations from '../../redux/mainCryptocurrency/mainCryptoOperations';
import mainCryptoSelectors from '../../redux/mainCryptocurrency/mainCryptoSelectors';
import favCryptoSelectors from '../../redux/favCryptocurrency/favCryptoSelectors';
import MainListItem from '../MainListItem/MainListItem';
import withTitleList from '../../hoc/withTitleList';

const hocOptions = { title: 'Top cryptorurrencies:' };

const MainList = () => {
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(mainCryptoOperations.getAllCryptoCurrenciesOperation());
  }, [dispatch]);

  const allCryptoCurrencies = useSelector(
    mainCryptoSelectors.getAllCryptocurrencies,
  );

  const favCryptocurrencies = useSelector(
    favCryptoSelectors.getFavCryptocurrencies,
  );

  const cryptocurrenciesList = useMemo(() => {
    return (
      allCryptoCurrencies.length !== 0 &&
      allCryptoCurrencies.map(crypto => {
        const isFavorite = !!favCryptocurrencies.find(
          favCrypto => favCrypto.id === crypto.id,
        );
        return (
          <MainListItem
            key={crypto.id}
            name={crypto.name}
            symbol={crypto.symbol}
            price={crypto.price}
            id={crypto.id}
            isFavorite={isFavorite}
          />
        );
      })
    );
  }, [allCryptoCurrencies, favCryptocurrencies]);

  return (
    <ul>
      {allCryptoCurrencies.length !== 0
        ? cryptocurrenciesList
        : 'no results...'}
    </ul>
  );
};

export default withTitleList(hocOptions)(MainList);
