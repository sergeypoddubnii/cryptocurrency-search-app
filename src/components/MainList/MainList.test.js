import React from 'react';
import { shallow } from 'enzyme';
import MainList from './MainList.tsx';

describe('MainList', () => {
	let mainList;
	beforeEach(() => {
		// eslint-disable-next-line no-undef
		mainList = shallow(<MainList />);
	});
	it('should render correct', () => {
		expect(mainList).toMatchSnapshot();
	});
});
