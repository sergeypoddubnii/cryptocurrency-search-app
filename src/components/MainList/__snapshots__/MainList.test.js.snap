// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainList should render correct 1`] = `
<styled.div>
  <styled.h2>
    Top cryptorurrencies:
  </styled.h2>
  <MainList />
</styled.div>
`;
